text/0249-v2-experiments.md

```diff
@@ -0,0 +1,127 @@
+  documentation to point users to the new stable version in the main repository. The lifecycle of experimental modules
+  in the preview repository will be exactly the same as it is currently in 1.x, with the same metadata markers in the
+  code.
+- The contributor guide for the CDK will be modified so that users know where to introduce new experimental code.
+- If it becomes absolutely necessary to add a new feature to v2 that would be a breaking change for users, the feature
+  will be hidden behind a feature flag. The code path will not be accessible without a specific opt-in by the user.
+
+# Drawbacks
+
+- In several places in the current 1.x CDK branch, stable modules refer to or contain experimental code. This will no
+  longer be possible after the release of v2.
+- When we graduate a module, users who have participated in the Developer Preview will have to update their imports to
+  make use of the stable version in the main repository.
+- For developers who are working on experimental code that has complex dependencies with the core and with other
+  modules, development will be more complicated.
+- In some cases, experimentation will not be possible. New features introduced to the main repository will have to be
+  vetted much more carefully before they are merged, since any public API released as part of v2 will have to be
+  supported for the lifetime of the major version.
+
+# Alternatives
+
+Some alternatives to this strategy are:
+
+- Make no changes and continue to experiment in the main release. Improve documentation so that it is more obvious to
+  users that they are importing experimental code. We might also make it a policy to actually include the name
+  'experimental' in all experimental APIs.
+- Use clever build tricks to hide experimental modules. This might be possible in Typescript, where the type files are
+  separate from implementation, but might prove difficult or impossible in other languages such as Python.
+
+# Adoption Strategy
+
+The hope is that for the majority of users, upgrading from 1.x to 2.x will be as simple as changing import statements at
+the top of their source files, rebuilding, and redeploying. This upgrade should not result in any resources being
+destroyed and recreated, except in very specific cases where we have made it clear to users, such as with the EKS L2
+construct library. The same should be true for users upgrading from the preview repository to the stable repository as
+modules are graduated from Developer Preview to GA.
+
+# Unresolved questions
+
+- One debated decision was related to the preview repository. Should it be a single repository or should each new module
+  get its own repository? Separate repositories are attractive because it gives us a way to demonstrate a canonical way
+  to develop an independent construct library. But using a single repository is much simpler logistically.
```

Comment:
In my opinion, multiple repositories will make it effectively impossible for the experimental modules to depend on each other, because of dependency problems. This is a huge problem in my opinion.

Reply:
Isn't that by definition an L3, though, when we write a construct that has dependencies on several other libraries? We have talked about limiting aws-cdk-lib to L2s, which should be for the most part focused on a single service, and moving L3s out to their own repositories.

Reply:
I don't follow... [`@aws-cdk/aws-ecs` has 23 dependencies](https://github.com/aws/aws-cdk/blob/857acbbb7f26feecca938dc881add57fe5cae7e4/packages/%40aws-cdk/aws-ecs/package.json#L78-L100), does that mean it's not an L2, but an L3, and we should move it out of mono-CDK...?

Reply:
Also, I don't think the numbers matter here. The moment you have even a single dependency between two separate experimental modules, you introduce this problem (inability to upgrade the CDK version) for NPM consumers.